cluster-ui: move table details endpoint to use sql-over-http

THardy98 opened the pull request:
Part of: https://github.com/cockroachdb/cockroach/issues/89429
Resolves: https://github.com/cockroachdb/cockroach/issues/90264, https://github.com/cockroachdb/cockroach/issues/90268

This PR migrates our existing admin API table details endpoint to
use sql-over-http in cluster-ui. We change the structure of the table
details response, allowing each field to be a self-contained query
response. The exception here is the `stats` field, which is comprised of
a couple query responses. Changing the structure scopes errors at the
query-level, instead of the response level. This is important as it
allows us to still make use of partial data on the UI when only a subset
of the queries succeed.

The commit also coalesces the former table stats and table details API
into a single table details API. The only difference/movement here is
that the table details endpoint now fetches span statistics.

The data for table details is populated using 10 queries, which fetch
the table's:
- ID
- grants
- schema details (columns & indexes)
- create statement
- zone configuration statement
- zone configuration
- heuristics details (specifically, the timestamp at which table
  heuristics were last updated)
- span statistics
- index usage statistics
- replicas

This largely reflects like for like how we fetch data from the existing
admin api endpoint. We can improve upon this (i.e. remove bloating) in
follow up PRs.

**Short Demo**
https://www.loom.com/share/698c10c5752e439a9edb0ec2301771c1

Release note: None

Commit:
cluster-ui: migrate table details api to use sql-over-http

This commit migrates our existing admin API table details endpoint to
use sql-over-http in cluster-ui. We change the structure of the table
details response, allowing each field to be a self-contained query
response. The exception here is the `stats` field, which is comprised of
a couple query responses. Changing the structure scopes errors at the
query-level, instead of the response level. This is important as it
allows us to still make use of partial data on the UI when only a subset
of the queries succeed.

The commit also coalesces the former table stats and table details API
into a single table details API. The only difference/movement here is
that the table details endpoint now fetches span statistics.

The data for table details is populated using 10 queries, which fetch
the table's:
- ID
- grants
- schema details (columns & indexes)
- create statement
- zone configuration statement
- zone configuration
- heuristics details (specifically, the timestamp at which table
  heuristics were last updated)
- span statistics
- index usage statistics
- replicas

This largely reflects like for like how we fetch data from the existing
admin api endpoint. We can improve upon this (i.e. remove bloating) in
follow up PRs.

Release note: None

https://cockroachlabs.atlassian.net/browse/DOC-1355 #Informs:
https://github.com/cockroachdb/cockroach/issues/33316 #Epic: CRDB-8035

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98326)
<!-- Reviewable:end -->

j82w reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98326)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @THardy98)

___
*[`pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts` line 289 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-NQAjeQZEcvyQeh5dwOz:-NQAjeQZEcvyQeh5dwP-:b4q5uc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d098b479cb7fa66b8b8da6564b75b7aaa0351a65/pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts#L289)):*
> ```XML
>       // and not empty.
>       if (
>         txn_result.rows[0].database_zone_config_bytes &&
> ```

nit: ` txn_result.rows[0]` seems to be accessed a lot in this method. Why not set it to a const and use that everywhere to avoid the long lines of code?
___
*[`pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts` line 292 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-NQAiuxdE2zC1FtltPWE:-NQAiuxdE2zC1FtltPWF:bq72oj5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d098b479cb7fa66b8b8da6564b75b7aaa0351a65/pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts#L292)):*
> ```XML
>         txn_result.rows[0].database_zone_config_bytes.length !== 0
>       ) {
>         zoneConfigData.hexString =
> ```

Any way to use the sql function `encode(column, 'hex')` to remove the hex logic?
___
*[`pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts` line 333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-NQAiSDcDrH-2cZcegW7:-NQAiSDcDrH-2cZcegW8:bw1ta0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/d098b479cb7fa66b8b8da6564b75b7aaa0351a65/pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts#L333)):*
> ```XML
> 
> // Table heuristics details.
> type TableHeuristicsDetails = TableHeuristicDetailsRow & {
> ```

Any reason you aren't using the new SqlApiResponse for all these types?
___
*[`pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts` line 420 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-NQAkVyY70NmW6DG1dPc:-NQAkVyY70NmW6DG1dPd:b-jdripp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d098b479cb7fa66b8b8da6564b75b7aaa0351a65/pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts#L420)):*
> ```XML
>     if (!(!txn_result.rows || txn_result.rows?.length === 0)) {
>       resp.stats.pebble_data.approximate_disk_bytes =
>         txn_result.rows[0].approximate_disk_bytes;
> ```

Is there a guarantee that it will always be 1 row? If it is please add a check and log to console if there is more than 1 row. It's very hard to identify and root cause partial results.
___
*[`pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts` line 518 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-NQAmElw06xRUrSg6mui:-NQAmElw06xRUrSg6muj:b8xdizt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d098b479cb7fa66b8b8da6564b75b7aaa0351a65/pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts#L518)):*
> ```XML
>         [new Identifier(dbName)],
>       ),
>       arguments: [escFullTableName.SQLString(), dbName],
> ```

Should we get both the name and ids for the dbs/tables to avoid the overhead of the lookup?
___
*[`pkg/ui/workspaces/cluster-ui/src/api/util.ts` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-NQAhp7C6soRJ2wyT5Zg:-NQAhp7C6soRJ2wyT5Zh:bwgcik6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d098b479cb7fa66b8b8da6564b75b7aaa0351a65/pkg/ui/workspaces/cluster-ui/src/api/util.ts#L44)):*
> ```XML
> }
> 
> export const stripLeadingHexMarker = (hexString: string): string => {
> ```

Should FixFingerprintHexValue be used instead?


<!-- Sent from Reviewable.io -->

Commit:
ui: update databaseDetailsPage

Update the database details page to reflect the new response structure.

Release note: None

https://cockroachlabs.atlassian.net/browse/DOC-1355 #Informs:
https://github.com/cockroachdb/cockroach/issues/33316 #Epic: CRDB-8035

THardy98 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98326)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @j82w)

___
*[`pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts` line 289 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-NQAjeQZEcvyQeh5dwOz:-NQfNHxK-6hu2IPafCKI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d098b479cb7fa66b8b8da6564b75b7aaa0351a65/pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts#L289)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

nit: ` txn_result.rows[0]` seems to be accessed a lot in this method. Why not set it to a const and use that everywhere to avoid the long lines of code?
</blockquote></details>

Done.
___
*[`pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts` line 292 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-NQAiuxdE2zC1FtltPWE:-NQfNJCaE-u0UCHE7AnM:bha03ib) ([raw file](https://github.com/cockroachdb/cockroach/blob/d098b479cb7fa66b8b8da6564b75b7aaa0351a65/pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts#L292)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

Any way to use the sql function `encode(column, 'hex')` to remove the hex logic?
</blockquote></details>

Done, able to remove some but not all.
___
*[`pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts` line 333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-NQAiSDcDrH-2cZcegW7:-NQfNOECARKJFNu4VF5D:b-1p0r7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d098b479cb7fa66b8b8da6564b75b7aaa0351a65/pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts#L333)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

Any reason you aren't using the new SqlApiResponse for all these types?
</blockquote></details>

Discussed offline, opted to use `SqlApiQueryResponse` to capture query-level errors.
___
*[`pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts` line 420 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-NQAkVyY70NmW6DG1dPc:-NQfNVS1Bcx3uaUumR1S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d098b479cb7fa66b8b8da6564b75b7aaa0351a65/pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts#L420)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

Is there a guarantee that it will always be 1 row? If it is please add a check and log to console if there is more than 1 row. It's very hard to identify and root cause partial results.
</blockquote></details>

Done.
___
*[`pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts` line 518 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-NQAmElw06xRUrSg6mui:-NQeWFV4CKOFWhIXDB2U:be0fui9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d098b479cb7fa66b8b8da6564b75b7aaa0351a65/pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts#L518)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

Should we get both the name and ids for the dbs/tables to avoid the overhead of the lookup?
</blockquote></details>

I'd like to do this in a follow-up PR that we could backport. Personally, I'd like to get names and ids for databases in the `databasesApi.ts` and introduce a similar api for tables `tablesListApi.ts` to get table names/ids.
___
*[`pkg/ui/workspaces/cluster-ui/src/api/util.ts` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-NQAhp7C6soRJ2wyT5Zg:-NQfNDXDDg_QL8gdXGr3:b-ps0t29) ([raw file](https://github.com/cockroachdb/cockroach/blob/d098b479cb7fa66b8b8da6564b75b7aaa0351a65/pkg/ui/workspaces/cluster-ui/src/api/util.ts#L44)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

Should FixFingerprintHexValue be used instead?
</blockquote></details>

Removed this function, using `encode(..., 'hex')`


<!-- Sent from Reviewable.io -->

maryliag reviewed (approved):
Great work on this!! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 16 of 16 files at r4, 6 of 6 files at r5, 8 of 8 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-:-NQkFJ2k629tNIH3Lu7D:bpkt4nw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @j82w)



<!-- Sent from Reviewable.io -->

j82w reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 16 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-:-NQkHSiX6GHQ1y2_ocuv:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @THardy98)

___
*[`pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts` line 137 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98326#-NQjup72EC0XmfrvK4bv:-NQjup72EC0XmfrvK4bw:b-85iey1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd0898562eb2a77ad783c6f06c6bc9b766479030/pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts#L137)):*
> ```XML
>       );
>     }
>     if (txn_result.error) {
> ```

nit: if is unnecessary 


<!-- Sent from Reviewable.io -->

xinhaoz commented on pkg/ui/workspaces/db-console/src/views/databases/databaseDetailsPage/redux.ts:
```diff
@@ -141,51 +133,48 @@ export const mapStateToProps = createSelector(
           ).length;
           return {
             name: table,
+            loading: !!details?.inFlight,
```
Is there a ticket to change this mapStateFromProps with some selectors on fields to a regular function rather than a selector?

THardy98 replied:
I don't believe so.

xinhaoz commented on pkg/ui/workspaces/db-console/src/views/databases/databaseTablePage/redux.ts:
```diff
@@ -76,7 +72,6 @@ export const mapStateToProps = createSelector(
   ): DatabaseTablePageData => {
     const details = tableDetails[generateTableID(database, table)];
-    const stats = tableStats[generateTableID(database, table)];
     const indexStats = indexUsageStats[generateTableID(database, table)];
```
Same with this mapStateToProps

THardy98 replied:
Ditto.

Commit:
ui: update databaseTablePage
Update the database table page to reflect the new response structure.

Release note: None

THardy98 commented on pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts:
```diff
@@ -0,0 +1,612 @@
+  };
+}
+
+export type TableDetailsResponse = {
```
Changed the top level fields to use camelCase. Inner fields follow the same format that's being returned by the sql api which is snake case.

THardy98 commented on pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts:
```diff
@@ -0,0 +1,612 @@
+// Table schema details.
+type TableSchemaDetails = TableSchemaDetailsRow & {
+  error?: Error;
+};
```
Done.

THardy98 commented:
bors r+

maryliag commented:
bors r+
